<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

---
# Exercise 1

In [ ]:
odd_list = [] # initialise list
for number in range(9):
    if number%2 == 0:
        continue # if even, skips to next loop
    odd_list.append(number) # the number is odd, add to list

print(odd_list)

[1, 3, 5, 7]


---
# Exercise 2

In [ ]:
# list comprehension: only add to list if odd
[number for number in range(9) if number%2 == 1] 

[1, 3, 5, 7]

---
# Exercise 3

In [4]:
import timeit

ex1 = '''odd_list = []
for number in range(9):
    if number%2 == 0:
        continue
    odd_list.append(number)'''
ex2 = '''[number for number in range(9) if number%2 == 1]'''

print(timeit.timeit(ex1))
print(timeit.timeit(ex2))


0.40922724999836646
0.35159837498213165


## Result:
Time taken for Exercise 1: 0.40922724999836646 \
Time taken for Exercise 2: 0.35159837498213165

---
# Exercise 4

In [ ]:
year_counter = 0
ini_units = 100

while True: #keep looping
    ini_units = ini_units * 0.95 #how much left in current year
    year_counter += 1 #current year
    if ini_units <= 50:
        break #stop

print(f'The sample took {year_counter} years to halve in amount.')

The sample took 14 years to halve in amount.


## Result: 
The sample took 14 years to halve in amount.

---
# Exercise 5

In [52]:
import numpy as np
co2_data = np.array([
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
])

co2_array = co2_data[0:(co2_data.size), 1] # making array of numbers
co2_diff = np.diff(co2_array) # find differences
co2_diff = np.insert(co2_diff,0,0) # make first entry 0, to match shape for mask
#print(co2_diff)

diff_mask = co2_diff >= 3 # make mask

big_diff = co2_data[diff_mask] # mask main one
co2_diff = co2_diff[diff_mask] # mask diff
print(big_diff)
print(f'Year: {int(big_diff[0, 0])}, change: {co2_diff[0]}')

[[2016.    404.24]]
Year: 2016, change: 3.410000000000025
